File: ollama-proxy.js
#!/usr/bin/env node
/**
 * Simple proxy server that adds CORS headers to Ollama requests
 * This allows the browser to talk to Ollama without CORS issues
 */

import express from 'express';
import cors from 'cors';

const app = express();
const PORT = 5105;
const OLLAMA_URL = 'http://localhost:11434';

app.use(cors());
app.use(express.json());

// Health check
app.get('/health', (req, res) => {
  res.json({ status: 'ok', service: 'Ollama Proxy' });
});

// Proxy all requests to Ollama
app.all('/api/*', async (req, res) => {
  try {
    const ollamaPath = req.path.replace('/api', '');
    const ollamaEndpoint = `${OLLAMA_URL}${ollamaPath}`;

    console.log(`[Proxy] ${req.method} ${ollamaEndpoint}`);

    const response = await fetch(ollamaEndpoint, {
      method: req.method,
      headers: {
        'Content-Type': 'application/json',
      },
      body: req.method !== 'GET' ? JSON.stringify(req.body) : undefined
    });

    const data = await response.json();
    res.json(data);
  } catch (error) {
    console.error('[Proxy] Error:', error.message);
    res.status(500).json({ error: error.message });
  }
});

app.listen(PORT, () => {
  console.log(`✅ Ollama Proxy running on http://localhost:${PORT}`);
  console.log(`✅ Forwarding requests to: ${OLLAMA_URL}`);
  console.log(`✅ CORS enabled for browser access`);
});

File: ollama-proxy.py
#!/usr/bin/env python3
"""
Simple proxy server that adds CORS headers to Ollama requests
This allows the browser to talk to Ollama without CORS issues
"""
from http.server import HTTPServer, BaseHTTPRequestHandler
import json
import urllib.request
import urllib.error

OLLAMA_URL = 'http://localhost:11434'
PORT = 5105

class OllamaProxyHandler(BaseHTTPRequestHandler):
    def _send_cors_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')

    def do_OPTIONS(self):
        self.send_response(200)
        self._send_cors_headers()
        self.end_headers()

    def do_GET(self):
        if self.path == '/health':
            self.send_response(200)
            self._send_cors_headers()
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            response = json.dumps({'status': 'ok', 'service': 'Ollama Proxy'})
            self.wfile.write(response.encode())
            return

        # Forward to Ollama
        self._proxy_request()

    def do_POST(self):
        self._proxy_request()

    def _proxy_request(self):
        try:
            # Build Ollama URL - forward the path as-is
            ollama_endpoint = f"{OLLAMA_URL}{self.path}"

            print(f"[Proxy] {self.command} {ollama_endpoint}")

            # Read request body for POST
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length) if content_length > 0 else None

            # Make request to Ollama
            req = urllib.request.Request(
                ollama_endpoint,
                data=body,
                headers={'Content-Type': 'application/json'}
            )

            with urllib.request.urlopen(req, timeout=60) as response:
                data = response.read()

                self.send_response(200)
                self._send_cors_headers()
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(data)

        except urllib.error.URLError as e:
            print(f"[Proxy] Error: {e}")
            self.send_response(500)
            self._send_cors_headers()
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            error = json.dumps({'error': str(e)})
            self.wfile.write(error.encode())

    def log_message(self, format, *args):
        # Custom log format
        pass

def run_proxy(port=5105):
    server_address = ('', port)
    httpd = HTTPServer(server_address, OllamaProxyHandler)
    print(f'✅ Ollama Proxy running on http://localhost:{port}')
    print(f'✅ Forwarding requests to: {OLLAMA_URL}')
    print(f'✅ CORS enabled for browser access')
    print('---')
    httpd.serve_forever()

if __name__ == '__main__':
    run_proxy()
